### PR TITLE
Proto API for LB request/responses

### PR DESCRIPTION
**Pull request by dgquintas**
NB: we may want to move the nanopb files (nanopb/pb_*) into a submodule.

[Review on Reviewable](https://reviewable.io/reviews/grpc/grpc/4038)

### CODE REVIEW

**Comment by ctiller** [comment]
This needs a script in tools/codegen/core to regenerate the nanopb generated files to the right place

File: src/core/client_config/lb_policies/protos/load_balancer.proto [hunk 1]
```diff
@@ -0,0 +1,115 @@
+syntax = "proto3";
```
Comment (ctiller) [inline src/core/client_config/lb_policies/protos/load_balancer.proto]: @a11r, @ejona86: where's the current best-practice place to locate .proto files?

Comment (ejona86) [inline src/core/client_config/lb_policies/protos/load_balancer.proto]: Isn't it different for each repo? I'd imagine a src/proto mirroring test/proto would be reasonable, and easy for us in Java to find.

**Comment by nicolasnoble** [comment]
For the sake of distinguishing what's what, I'll agree with Craig: we probably should move anything that we didn't write ourselves into the third_party folder. Even if it's not a submodule.

**Comment by dgquintas** [comment]
BUMP. Updated.

**Comment by dgquintas** [comment]
Bump. Finally "green".

File: src/core/client_config/lb_policies/proto/load_balancer.proto [hunk 2]
```diff
@@ -0,0 +1,115 @@
+syntax = "proto3";
```
Comment (ctiller) [inline src/core/client_config/lb_policies/proto/load_balancer.proto]: This should be in src/proto/grpc/lb/v0/...

File: BUILD [hunk 3]
```diff
@@ -226,6 +227,7 @@ cc_library(
     "src/core/json/json_common.h",
     "src/core/json/json_reader.h",
     "src/core/json/json_writer.h",
+    "src/core/proto/load_balancer.pb.h",
```
Comment (ctiller) [inline BUILD]: And then this should be in the same directory as the .proto

**Comment by ctiller** [comment]
Can we get a script that runs the proto compilation process as part of this, and a script added to run_sanity.sh that verifies the generated code is up to date?

File: BUILD (hunk 3 above)
Comment (dgquintas) [inline BUILD]: The problem with making the generated files go there is that the nanopb generated pb.{c,h} will clash with the ones generated by the regular C++ proto compiler. One or the other ought to output their generated code to a non-standard (ie, not following the proto package hierarchy) directory.

Comment (dgquintas) [inline BUILD]: The C++ output is used in the grpclb_api_test file to encode/decode the test protos used, for convenience (the C api doesn't have all the encoding/decoding functions, only the ones needed by the grpclb policy).

Comment (ctiller) [inline BUILD]: Hmm. I'm against scattering proto related stuff all over the tree. How do you feel about src/core/proto/.../xxx.pb.h for the nanopb generated stuff for src/proto/.../xxx.proto?

Comment (dgquintas) [inline BUILD]: That's precisely what we have. The problem found while introducing the change you proposed is that now the nanopb outputs xxx.pb.h to .../grpc/lb/v0, the same place the C++ proto compiler will (by virtue of the "package grpc.lb.v0" line in the proto). The C++ .h will then wreak havoc in all the C files that include and expect nanopb's C pb.h.

We either stop relying on the C++ generated code for the .proto in the tests (bummer, will have to augment the C LB proto API just for testing) or we output the nanopb generated code to a non-package-consistent directory (as before).

**Comment by dgquintas** [comment]
Added sanity scripts and checks... waiting to see how they fare inside the Docker container.

**Comment by dgquintas** [comment]
Sanity finally passes when run locally with run_tests.py -l sanity --use-docker. HOWEVER, docker seems to be having a rough day today... merge at will.

**Comment by dgquintas** [comment]
WOOHOO, finally green for sanity!

**Comment by dgquintas** [comment]
Ping, before something else breaks again :)

**Pull request merged**